include/swift/Basic/CXXStdlibKind.h

```diff
@@ -0,0 +1,30 @@
+//===--- CXXStdlibKind.h ----------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_CXX_STDLIB_KIND_H
+#define SWIFT_BASIC_CXX_STDLIB_KIND_H
+
+namespace swift {
+
+enum class CXXStdlibKind : uint8_t {
+  /// Use the default C++ stdlib for a particular platform: libc++ for Darwin,
+  /// libstdc++ for most Linux distros.
+  PlatformDefault = 0,
```

Comment:
I'm not a huge fan of embedding "platform default" vs "libcxx" in the compiler given that our notion of "default" isn't necessarily right. Android is technically a Linux, where the NDK provides a libcxx to link against. But, there isn't a standard system-wide C++ runtime either since Google doesn't guarantee C++ ABI stability, it's whatever was shipped in the NDK. That also means that I could totally use libstdc++ with my program if I wanted to and neither is the default.

should this enum instead list out the supported C++ runtimes (which are enumerable since we also have to provide the modulemap bits anyway). Then it's less prescriptive, and we could even consider exposing a mechanism for the vendor and the SDK to say what their "default" C++ runtime is.

Reply:
I think a reasonable definition of the "default" here would be: what library Clang picks when you run `clang++ x.cpp`. Having a different default than Clang would be quite disastrous since people would try to build a C++ library with Clang (without passing `-stdlib=xyz`) and then use it from Swift (again, without passing `-Xcc -stdlib=xyz`) and not realize that they're using two different C++ stdlibs.

On Android, Clang uses libc++ by default, and `PlatformDefault` would mean libc++ when targeting Android.

The actual meaning of `PlatformDefault` here is that the library on client's side is the same as the library that Swift was built against, meaning the compiler is free to use prebuilt binaries such as the CxxStdlib overlays.

Reply:
Clang now uses libc++ by default, but if you change the platform version, the default should be gnustl (aka libstdc++). They also recommended stlport for a while. I think that "platform default" just doesn't really give you much information (especially as a user).

Reply:
From what I can tell, Clang seems to use libc++ as the default on Android regardless of the version: https://github.com/llvm/llvm-project/blob/bc747c3e1377e3e262d901cefcae2f563c895bb5/clang/lib/Driver/ToolChains/Linux.cpp#L367-L368